### PR TITLE
i3s: load draco geometry

### PR DESCRIPTION
**Pull request by belom88**

### CODE REVIEW

File: examples/website/i3s/examples.js [hunk 1]
```diff
@@ -22,6 +22,16 @@ export const EXAMPLES = {
       latitude: 34
     }
   },
+  'San Francisco v1.7': {
```
Comment (ibgreen) [inline examples/website/i3s/examples.js:25]: Should we mark each example with the version?

File: modules/i3s/docs/api-reference/i3s-loader.md [hunk 2]
```diff
@@ -159,15 +159,14 @@ const visibleTiles = tileset3d.tiles.filter(tile => tile.selected);
...
+| `options.i3s.tileset`          | `Object`         | `null`  | `Tileset` object loaded by I3SLoader or follow the data format specified in [Tileset Object](#tileset-object). It is required when loading i3s geometry content                                                                                                                                                                 |
+| `options.i3s.tile`             | `Object`         | `null`  | `Tile` object loaded by I3SLoader or follow the data format [Tile Object](#tile-object). It is required when loading i3s geometry content                                                                                                                                                                                       |
+| `options.i3s.useDracoGeometry` | `Bool`           | `false` | Use 'Draco' compressed geometry to show if applicable                                                                                                                                                                                                                                                                           |
```
Comment (ibgreen) [inline modules/i3s/docs/api-reference/i3s-loader.md:169]: Don't we want this to be true to by default?

File: modules/i3s/src/helpers/i3s-nodepages-tiles.js [hunk 3]
```diff
@@ -48,11 +48,13 @@ export default class I3SNodePagesTiles {
     let materialDefinition = null;
     let textureFormat = 'jpeg';
     let attributeUrls = [];
+    let isDracoGeometry = false;
 
     if (node && node.mesh) {
-      if (node.mesh.geometry) {
-        contentUrl = `${this.tileset.url}/nodes/${node.mesh.geometry.resource}/geometries/0`;
-      }
+      const {url, isDracoGeometry: isDracoGeometryResult} = (node.mesh.geometry &&
```
Comment (ibgreen) [inline modules/i3s/src/helpers/i3s-nodepages-tiles.js:54]: This looks like the wrong place to know about whether things are draco encoded or not. Can't this be handled and results "normalized" on the node level?

The expressions are a bit messy, if possible make more effort to write this cleanly.

Please add some comment.

File: modules/i3s/src/helpers/i3s-nodepages-tiles.js [hunk 4]
```diff
@@ -92,8 +84,50 @@ export default class I3SNodePagesTiles {
...
+    if (this.options.i3s && this.options.i3s.useDracoGeometry) {
+      geometryIndex = geometryDefinition.geometryBuffers.findIndex(
+        buffer => buffer.compressedAttributes && buffer.compressedAttributes.encoding === 'draco'
+      );
+    }
+    if (geometryIndex === -1) {
+      geometryIndex = geometryDefinition.geometryBuffers.findIndex(
+        buffer => !buffer.compressedAttributes
+      );
+    }
```
Comment (ibgreen) [inline modules/i3s/src/helpers/i3s-nodepages-tiles.js:105]: Use `else` ?

File: modules/i3s/src/helpers/i3s-nodepages-tiles.js [hunk 5]
```diff
@@ -92,8 +84,50 @@ export default class I3SNodePagesTiles {
...
+      );
+      result = {
+        url: `${this.tileset.url}/nodes/${meshGeometryData.resource}/geometries/${geometryIndex}`,
+        isDracoGeometry
+      };
+    }
+    return result;
+  }
+
+  _getLodSelection(node) {
```
Comment (ibgreen) [inline modules/i3s/src/helpers/i3s-nodepages-tiles.js:118]: JSDoc and types everywhere please.

File: modules/i3s/src/helpers/i3s-nodepages-tiles.js [hunk 6]
```diff
@@ -92,8 +84,50 @@ export default class I3SNodePagesTiles {
       attributeUrls,
       materialDefinition,
       textureFormat,
-      children
+      children,
+      isDracoGeometry
+    });
+  }
+
+  _getContentUrl(meshGeometryData) {
```
Comment (ibgreen) [inline modules/i3s/src/helpers/i3s-nodepages-tiles.js:92]: JSDoc

File: modules/i3s/src/helpers/i3s-nodepages-tiles.js [hunk 7]
```diff
@@ -92,8 +84,50 @@ export default class I3SNodePagesTiles {
...
+    let result = {};
+    const geometryDefinition = this.tileset.geometryDefinitions[meshGeometryData.definition];
+    let geometryIndex = -1;
+    if (this.options.i3s && this.options.i3s.useDracoGeometry) {
```
Comment (ibgreen) [inline modules/i3s/src/helpers/i3s-nodepages-tiles.js:96]: Short comment explaining what this code is doing?

File: modules/i3s/src/lib/parsers/parse-i3s-tile-content.js [hunk 8]
```diff
@@ -121,6 +121,12 @@ async function parseI3SNodeGeometry(arrayBuffer, tile = {}, options) {
     faceRange: attributes.faceRange
   };
 
+  for (const attributeIndex in content.attributes) {
```
Comment (ibgreen) [inline modules/i3s/src/lib/parsers/parse-i3s-tile-content.js:124]: one line comment?

File: modules/i3s/src/lib/parsers/parse-i3s-tile-content.js [hunk 9]
```diff
@@ -279,11 +289,13 @@ function parsePositions(attribute, tile) {
   };
 }
 
-function offsetsToCartesians(vertices, cartographicOrigin) {
+function offsetsToCartesians(vertices, metadata, cartographicOrigin) {
```
Comment (ibgreen) [inline modules/i3s/src/lib/parsers/parse-i3s-tile-content.js:292]: JSDoc with types

**Review by ibgreen — APPROVED** [review]

**Comment by ibgreen** [comment]
Is there anyway we can roughly measure time to load a certain number of tiles before and after this PR?

**Comment by belom88** [comment]
> Is there anyway we can roughly measure time to load a certain number of tiles before and after this PR?

We have "Add a benchmark to ensure that performance does not degrade." in the work scope.

File: modules/i3s/docs/api-reference/i3s-loader.md (hunk 2 above)
Comment (belom88) [reply modules/i3s/docs/api-reference/i3s-loader.md:169]: Right

**Comment by belom88** [comment]
I'll fix comments in another PR

**Comment by ibgreen** [comment]
>  to ensure that performance does not degrade

Sure. the entire point of draco is to get a significant improvement because of improved loading speeds.

I suggest just printing elapsed time when tiles render for first time.

You can just use window.performance.now, but we should introduce probe.gl for this timing.

File: modules/i3s/src/helpers/i3s-nodepages-tiles.js (hunk 4 above)
Comment (belom88) [reply modules/i3s/src/helpers/i3s-nodepages-tiles.js:105]: No. After this `if` `geometryIndex` can get >=0

Comment (ibgreen) [reply modules/i3s/src/helpers/i3s-nodepages-tiles.js:105]: I see. A few short comments can help guide the reader.

**Pull request merged**